### PR TITLE
Validate Backbone Compatibility

### PR DESCRIPTION
**Pull request by Magnus-Kuhn**
# Readiness checklist

-   [x] I added/updated tests.
-   [x] I ensured that the PR title is good enough for the changelog.
-   [x] I labeled the PR.

### CODE REVIEW

File: packages/transport/src/core/Transport.ts [hunk 1]
```diff
@@ -59,6 +63,8 @@ export class Transport {
         allowIdentityCreation: true,
         supportedDatawalletVersion: 1,
         supportedIdentityVersion: -1,
+        supportedMinBackboneVersion: 5,
+        supportedMaxBackboneVersion: 5,
```
Comment (Magnus-Kuhn) [inline packages/transport/src/core/Transport.ts:67]: don't know what to put here

Comment (jkoenig134) [reply packages/transport/src/core/Transport.ts:67]: 5 seems fine, we don't know if we will be compatible with v6

File: packages/transport/test/modules/account/AccountController.test.ts
Comment (jkoenig134) [inline packages/transport/test/modules/account/AccountController.test.ts:1]: all these new tests will fail if the backbone reaches a specific major version. Wtb querying the backbone version before the test and adding / subtracting something from the actual version? Then the test is future proof.

File: packages/runtime/src/Runtime.ts [hunk 2]
```diff
@@ -191,6 +191,14 @@ export abstract class Runtime<TConfig extends RuntimeConfig = RuntimeConfig> {
         this._anonymousServices = Container.get<AnonymousServices>(AnonymousServices);
     }
 
+    public async validateUsedBackboneVersion(): Promise<Result<void>> {
```
Comment (jkoenig134) [inline packages/runtime/src/Runtime.ts:194]: "used" backbone version sounds quite off. WTB `checkBackboneCompatiblity`? This naming has to be aligned everywhere then.

File: packages/runtime/src/Runtime.ts [hunk 3]
```diff
@@ -191,6 +191,14 @@ export abstract class Runtime<TConfig extends RuntimeConfig = RuntimeConfig> {
...
+        if (!this._accountController) {
```
Comment (jkoenig134) [inline packages/runtime/src/Runtime.ts:195]: I don't see a reason why this is only possible with an active account.

File: packages/transport/src/core/Transport.ts [hunk 4]
```diff
@@ -36,6 +38,8 @@ export interface IConfigOverwrite {
     platformClientId: string;
     platformClientSecret: string;
     supportedIdentityVersion: number;
+    supportedMinBackboneVersion?: number;
+    supportedMaxBackboneVersion?: number;
```
Comment (jkoenig134) [inline packages/transport/src/core/Transport.ts:42]: this should not be overridable

File: packages/transport/src/core/Transport.ts (hunk 1 above)
Comment (jkoenig134) [inline packages/transport/src/core/Transport.ts:67]:
```suggestion
        supportedMinBackboneVersion: 6,
        supportedMaxBackboneVersion: 6,
```

File: packages/transport/test/modules/account/AccountController.test.ts
Comment (Magnus-Kuhn) [reply packages/transport/test/modules/account/AccountController.test.ts:1]: Just a number is simple, they only need to be large enough

File: packages/runtime/src/Runtime.ts (hunk 2 above)
Comment (Magnus-Kuhn) [reply packages/runtime/src/Runtime.ts:194]: Both are fine to me

File: packages/transport/src/core/Transport.ts (hunk 4 above)
Comment (Magnus-Kuhn) [reply packages/transport/src/core/Transport.ts:42]: I put the overrides in for testing - changing the limits is necessary for the test if there is no mock

File: packages/runtime/src/Runtime.ts (hunk 3 above)
Comment (jkoenig134) [reply packages/runtime/src/Runtime.ts:195]: will you change this? YOu thumbed up.

Comment (Magnus-Kuhn) [reply packages/runtime/src/Runtime.ts:195]: yeah, I will

File: packages/transport/test/modules/account/AccountController.test.ts
Comment (jkoenig134) [reply packages/transport/test/modules/account/AccountController.test.ts:1]: Why not make it really future proof bei just taking the actual version and add to it? Why just increase numbers..?

File: packages/transport/src/modules/accounts/AccountController.ts
Comment (jkoenig134) [inline packages/transport/src/modules/accounts/AccountController.ts:1]: I think this is not needed anymore

File: packages/runtime/src/extensibility/facades/anonymous/VersionFacade.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import { Result } from "@js-soft/ts-utils";
+import { Inject } from "typescript-ioc";
+import { CheckBackboneCompatibilityUseCase } from "../../../useCases/anonymous/CheckBackboneCompatibilityUseCase";
+
+export class VersionFacade {
```
Comment (jkoenig134) [inline packages/runtime/src/extensibility/facades/anonymous/VersionFacade.ts:5]: Version -> BackboneVersion or BackboneCompatibility

same for everywhere in this PR. I prefer BackboneCompatibility. The only place where the wording is valid is the VersionClient in transport.

File: packages/transport/test/modules/account/AccountController.test.ts
Comment (Magnus-Kuhn) [reply packages/transport/test/modules/account/AccountController.test.ts:1]: I'd have to think about how to get to the backbone version number and I don't expect enmeshed to still be around in the next millennium 💀

Comment (Milena-Czierlinski) [reply packages/transport/test/modules/account/AccountController.test.ts:1]: @jkoenig134 If we want to get the Backbone version, we should probably expose it via a method of the `BackboneCompatibilityController`. This might be a valid method exceeding just testing. Then, the question arises if `BackboneCompatibilityController` is still an appropriate name or if we should switch back to something more general.

(I also noticed that the wording `version` is still used in some places. Depending on the outcome of this conversion, I'll adjust it accordingly.)

Comment (jkoenig134) [reply packages/transport/test/modules/account/AccountController.test.ts:1]: > the next millennium

why do it hacky when you can do it clean?!

Comment (jkoenig134) [reply packages/transport/test/modules/account/AccountController.test.ts:1]: > If we want to get the Backbone version, we should probably expose it via a method of the BackboneCompatibilityController

we should not expose things only for testing.

Comment (Milena-Czierlinski) [reply packages/transport/test/modules/account/AccountController.test.ts:1]: > This might be a valid method exceeding just testing.

🤔

File: packages/runtime/src/useCases/anonymous/CheckBackboneCompatibilityUseCase.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import { Result } from "@js-soft/ts-utils";
+import { BackboneCompatibilityController } from "@nmshd/transport";
+import { Inject } from "typescript-ioc";
+import { UseCase } from "../common";
+
+export class CheckBackboneCompatibilityUseCase extends UseCase<void, void> {
```
Comment (jkoenig134) [inline packages/runtime/src/useCases/anonymous/CheckBackboneCompatibilityUseCase.ts:6]: void as successful result and throwing an error if it's not successful is very hard to process. Why not return a proper response that has 

1. a boolean to know if it's successful
2. an explanation on why it's not successful if it's not.

File: packages/transport/test/modules/backboneCompatibility/BackboneCompatibilityController.test.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import { IDatabaseConnection } from "@js-soft/docdb-access-abstractions";
+import { BackboneCompatibilityController, Transport } from "../../../src";
+import { TestUtil } from "../../testHelpers/TestUtil";
+
+describe("AccountController", function () {
```
Comment (jkoenig134) [inline packages/transport/test/modules/backboneCompatibility/BackboneCompatibilityController.test.ts:5]: naming

File: packages/transport/test/modules/account/AccountController.test.ts
Comment (jkoenig134) [reply packages/transport/test/modules/account/AccountController.test.ts:1]: missed this.. but what do we need this for?

Comment (Milena-Czierlinski) [reply packages/transport/test/modules/account/AccountController.test.ts:1]: Not sure at the current moment. But if we don't add a function for it, we'd have to make `client` of `BackboneCompatibilityController` public and I suppose we don't want that. Or do you see another way to get the Backbone version for testing?

Comment (jkoenig134) [reply packages/transport/test/modules/account/AccountController.test.ts:1]: it's js / ts ..  we don't have to make anything public to access it :D

Comment (jkoenig134) [reply packages/transport/test/modules/account/AccountController.test.ts:1]: I resolved it.

File: packages/runtime/src/extensibility/AnonymousServices.ts [hunk 8]
```diff
@@ -1,6 +1,9 @@
 import { Inject } from "typescript-ioc";
-import { AnonymousTokensFacade } from "./facades/anonymous";
+import { AnonymousTokensFacade, BackboneCompatibilityFacade } from "./facades/anonymous";
 
 export class AnonymousServices {
-    public constructor(@Inject public readonly tokens: AnonymousTokensFacade) {}
+    public constructor(
+        @Inject public readonly tokens: AnonymousTokensFacade,
+        @Inject public readonly version: BackboneCompatibilityFacade
```
Comment (Milena-Czierlinski) [inline packages/runtime/src/extensibility/AnonymousServices.ts:7]: `backboneCompatibility` instead of `version`.

File: packages/runtime/test/anonymous/version.test.ts
Comment (Milena-Czierlinski) [inline packages/runtime/test/anonymous/version.test.ts:1]: File naming

File: packages/runtime/test/anonymous/version.test.ts [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+afterEach(async () => {
+    await serviceProvider.stop();
+    await noLoginRuntime.stop();
+});
+
+describe("Backbone Compatibility Check", () => {
+    test("should successfully check the compatibility", async () => {
+        const transportConfigOverride = {
+            supportedMinBackboneVersion: 1,
+            supportedMaxBackboneVersion: 5000
```
Comment (Milena-Czierlinski) [inline packages/runtime/test/anonymous/version.test.ts:15]: Should we do some hacking here as well?

File: packages/transport/src/core/CoreErrors.ts [hunk 10]
```diff
@@ -197,6 +197,13 @@ class General {
     public invalidTruncatedReference() {
         return new CoreError("error.transport.files.invalidTruncatedReference", "invalid truncated reference");
     }
+
+    public runtimeVersionIncompatibleWithBackboneVersion(backboneVersion: number) {
+        return new CoreError(
+            "error.transport.files.runtimeVersionIncompatibleWithBackboneVersion",
```
Comment (Milena-Czierlinski) [inline packages/transport/src/core/CoreErrors.ts:203]: Should be `.general.` instead of `.files.`. Some other error codes here are inconsistent as well, but this is unrelated to the PR. Can we change it in a separate PR or would this be a breaking change such that we leave it for now?

File: packages/transport/test/modules/backboneCompatibility/BackboneCompatibilityController.test.ts [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+        const versionCheckResult = await versionController.checkBackboneCompatibility();
+        expect(versionCheckResult.isSuccess).toBe(true);
+    });
+
+    test("should catch a too low backbone version", async function () {
+        const versionController = new BackboneCompatibilityController({
+            ...transport.config,
+            supportedMinBackboneVersion: actualBackboneVersion + 1,
+            supportedMaxBackboneVersion: actualBackboneVersion + 2
```
Comment (Milena-Czierlinski) [inline packages/transport/test/modules/backboneCompatibility/BackboneCompatibilityController.test.ts:50]: I guess it's not important, but is there a reason for `+2` instead of `1` which is used in the other tests?

Comment (jkoenig134) [reply packages/transport/test/modules/backboneCompatibility/BackboneCompatibilityController.test.ts:50]: I think it doesn't matter here, if it's confusing I can use +1 instead.

File: packages/runtime/test/anonymous/version.test.ts (hunk 9 above)
Comment (jkoenig134) [reply packages/runtime/test/anonymous/version.test.ts]: yep :)

**Review by jkoenig134 — APPROVED** [review]

File: packages/runtime/src/useCases/anonymous/backboneCompatibility/CheckBackboneCompatibilityUseCase.ts [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+export interface CheckBackboneCompatibilityResponse {
+    isCompatible: boolean;
+    backboneVersion: number;
+    supportedMinBackboneVersion: number;
+    supportedMaxBackboneVersion: number;
+}
+
+export class CheckBackboneCompatibilityUseCase extends UseCase<void, CheckBackboneCompatibilityResponse> {
+    public constructor(@Inject private readonly anonymousVersionController: BackboneCompatibilityController) {
```
Comment (Milena-Czierlinski) [inline packages/runtime/src/useCases/anonymous/backboneCompatibility/CheckBackboneCompatibilityUseCase.ts:14]: `backboneCompatibilityController` instead of `anonymousVersionController`

**Review by Milena-Czierlinski — APPROVED** [review]

**Review by jkoenig134 — APPROVED** [review]
